### PR TITLE
add a katello plugin

### PR DESCRIPTION
**Pull request by cristifalcas**
This PR is adding the possibility to install and configure the katello plugin in foreman.

It will most probably fail if the foreman database is already populated. Currently it was only tested with upgrades.

Based on the discussion from issues/331 and mostly from the katello installation module, the order on which plugins and db scripts are executed was changed.

### CODE REVIEW

**Comment by ekohl** [comment]
Before we get into the code: would this make puppet-katello redundant? And do you mean https://github.com/theforeman/puppet-foreman/issues/331 as issue 331?

**Comment by cristifalcas** [comment]
This doesn't manage qpid-config and guterball installation

**Comment by ekohl** [comment]
So is the intention to ```include ::foreman::plugin::katello``` in puppet-katello? To be clear: I do think this could be a very good step, just trying to understand your goals.

**Comment by cristifalcas** [comment]
My goal is to manage katello together with foreman and separately from pulp/qpid/etc (which we have on other machines).

I noticed now that the original module also has some providers for katello. Should I copy them here?

File: manifests/init.pp [hunk 1]
```diff
@@ -2,193 +2,191 @@
 #
 # === Parameters:
 #
-# $foreman_url::                URL on which foreman is going to run
```
Comment (ekohl) [inline manifests/init.pp]: This appears to be unrelated. Maybe it needs a rebase?

File: manifests/init.pp [hunk 2]
```diff
@@ -201,7 +199,6 @@
   $passenger                 = $::foreman::params::passenger,
   $passenger_ruby            = $::foreman::params::passenger_ruby,
   $passenger_ruby_package    = $::foreman::params::passenger_ruby_package,
-  $plugin_prefix             = $::foreman::params::plugin_prefix,
```
Comment (ekohl) [inline manifests/init.pp]: This is a desired change that you're reverting.

File: manifests/plugin/katello.pp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  validate_bool($loggers_glue)
+  validate_bool($loggers_pulp_rest)
+  validate_bool($loggers_cp_rest)
+  validate_bool($loggers_cp_proxy)
+  validate_bool($loggers_action)
+  validate_bool($loggers_tire_rest)
+  validate_bool($loggers_manifest_import_logger)
+
+  if $configure_katello_repo {
+    case $::osfamily {
```
Comment (ekohl) [inline manifests/plugin/katello.pp]: Maybe this belongs in ```foreman::install::repos::katello```?

File: manifests/plugin/katello.pp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  include ::foreman::plugin::tasks
+
+  foreman::config::passenger::fragment { 'katello':
+    content     => template("${module_name}/plugins/katello_httpd.erb"),
+    ssl_content => template("${module_name}/plugins/katello_httpd.erb"),
+  }
+
+  foreman::plugin { 'katello':
+    package     => $package,
+    config_file => '/etc/foreman/plugins/katello.yaml',
```
Comment (ekohl) [inline manifests/plugin/katello.pp]: You can reuse ```$::foreman::plugin_config_dir```

File: templates/plugins/katello_httpd.erb [hunk 5]
```diff
@@ -0,0 +1,25 @@
+#
+# WARNING: THIS CONFIGURATION WAS GENERATED BY KATELLO-CONFIGURE TOOL,
+# CHANGES WILL LIKELY BE OVERWRITTEN.
+#
+
+SSLOptions +StdEnvVars +ExportCertData +FakeBasicAuth
+
+Alias /pub /var/www/html/pub
```
Comment (ekohl) [inline templates/plugins/katello_httpd.erb]: Are sure Katello needs this? I think this belongs to pulp.

File: templates/plugins/katello_httpd.erb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+<Location /pub>
+Options +FollowSymLinks +Indexes
+</Location>
+
+<LocationMatch /rhsm|/subscription|/katello/api>
```
Comment (ekohl) [inline templates/plugins/katello_httpd.erb]: Is rhsm and subscription part of the plugin? I thought that was a separate process which should configure its own apache.

**Comment by mmoll** [comment]
1. What's the status here?
2. Personally I'm unsure if this should go here, especially with the whole scenarios work and its background of making a unified installer for Foreman and Katello more likely

**Comment by ekohl** [comment]
Personally I'd like the case where Katello is just a plugin, but I think the rest of the ecosystem is not yet ready for this. I don't think it will be any time soon.

@ehelms what's your opinion on this?

**Comment by ehelms** [comment]
I don't see it in the near future either, until we handle a few other pieces and figure out a strategy for how to have Katello but without it's services directly available or available on another box as a first class citizen.

**Comment by ekohl** [comment]
When it's evolved enough, we should reconsider this but I'm going to close this for now.
